custom_components/localtuya/coordinator.py

```diff
@@ -556,77 +570,75 @@ def status_updated(self, status: dict):
     @callback
     def disconnected(self, exc=""):
         """Device disconnected."""
+        if not self._interface:
+            return
         self._interface = None
 
         if self._unsub_refresh:
             self._unsub_refresh()
+            self._unsub_refresh = None
 
-        if self.sub_devices:
-            for sub_dev in self.sub_devices.values():
-                sub_dev.disconnected("Gateway disconnected")
+        for subdevice in self.sub_devices.values():
+            subdevice.disconnected("Gateway disconnected")
 
-        if self._connect_task is not None:
-            self._connect_task.cancel("Device disconnected")
-            self._connect_task = None
+        if self._task_connect is not None:
+            self._task_connect.cancel()
+            self._task_connect = None
 
         # If it disconnects unexpectedly.
         if self._is_closing:
             return
 
-        self._call_on_close.append(asyncio.create_task(self._async_reconnect()).cancel)
-        self._call_on_close.append(
-            asyncio.create_task(self._shutdown_entities(exc=exc)).cancel
+        if self._task_reconnect is None:
+            self._task_reconnect = asyncio.create_task(self._async_reconnect())
+
+        if self._task_shutdown_entities is not None:
+            self._task_shutdown_entities.cancel()
+        self._task_shutdown_entities = asyncio.create_task(
+            self._shutdown_entities(exc=exc)
         )
 
     @callback
-    def subdevice_state(self, state: SubdeviceState):
+    def subdevice_state(self, state: SubdeviceState = None):
```

Comment:
Can you rename this callback to `subdevice_state_updated` 
and the state variable from `self._subdevice_state` to `self.subdevice_state`